projects/l2_pooling/cosyne_example.py

```diff
@@ -0,0 +1,168 @@
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""
+This file demonstrate convergence of single vs multiple columns
+"""
+
+import random
+import matplotlib.pyplot as plt
+import matplotlib as mpl
+
+from htmresearch.frameworks.layers.l2_l4_inference import L4L2Experiment
+from htmresearch.frameworks.layers.object_machine_factory import (
+  createObjectMachine
+)
+
+mpl.rcParams['pdf.fonttype'] = 42
+
+
+plt.ion()
+plt.close('all')
+
+
+def plotActivity(l2ActiveCellsMultiColumn):
+  fig, axs = plt.subplots(1, 3)
+  for c in range(len(l2ActiveCellsMultiColumn[0])):
+    for step in range(len(l2ActiveCellsMultiColumn)):
+      activeCellList = list(l2ActiveCellsMultiColumn[step][c])
+      for activeCells in activeCellList:
+        axs[c].plot(step, activeCells, '-ko')
+    axs[c].set_title('column {}'.format(c))
+    axs[c].set_xlabel('Time')
+  axs[0].set_ylabel('Neuron #')
+  return fig
+
+
+
+if __name__ == "__main__":
+  numColumns = [1, 3]
+  numFeatures = 3
+  numPoints = 10
+  numLocations = 10
+  numObjects = 10
+
+  objectMachine = createObjectMachine(
+    machineType="simple",
+    numInputBits=20,
+    sensorInputSize=1024,
+    externalInputSize=1024,
+    numCorticalColumns=3,
+    seed=37,
+  )
+  objectMachine.createRandomObjects(numObjects, numPoints=numPoints,
+                              numLocations=numLocations,
+                              numFeatures=numFeatures)
+
+
+  objects = objectMachine.provideObjectsToLearn()
+
+  # single-out the inputs to the column #1
+  objectsSingleColumn = {}
+  for i in range(numObjects):
+    featureLocations = []
+    for j in range(numLocations):
+      featureLocations.append({0: objects[i][j][0]})
+    objectsSingleColumn[i] = featureLocations
+
+  # we will run two experiments side by side, with either single column
+  # or 3 columns
+  exp3 = L4L2Experiment(
+    'three_column',
+    numCorticalColumns=3,
+    seed=1
+  )
+
+  exp1 = L4L2Experiment(
+    'single_column',
+    numCorticalColumns=1,
+    seed=1
+  )
+
+  print "train single column "
+  exp1.learnObjects(objectsSingleColumn)
+  print "train multi-column "
+  exp3.learnObjects(objects)
+
+  objectId = 0
+  obj = objectMachine[objectId]
+
+  # Create sequence of sensations for this object for all columns
+  numColumns = 3
+  objectSensations = {}
+  for c in range(numColumns):
+    objectCopy = [pair for pair in obj]
+    random.shuffle(objectCopy)
+    # stay multiple steps on each sensation
```

Comment:
Are we actually staying multiple steps on each sensation? I think we should, but I think this code is just putting each sensation in the list once. Is there a reason not to add these lines? https://github.com/numenta/nupic.research/blob/82ab5114841791f84f91c44ea0e83ed2390937cb/projects/l2_pooling/multi_column_convergence.py#L171-L173